crates/egui/src/atomic_layout.rs

```diff
@@ -0,0 +1,740 @@
+            kind: AtomicKind::Empty,
+        }
+    }
+
+    /// Heuristic to find the best height for an image.
+    /// Basically returns the height if this is not an [`Image`].
+    fn get_min_height_for_image(&self, fonts: &Fonts, style: &Style) -> Option<f32> {
+        self.size.map(|s| s.y).or_else(|| {
+            match &self.kind {
+                AtomicKind::Text(text) => Some(text.font_height(fonts, style)),
+                AtomicKind::Custom(_, size) => Some(size.y),
+                // Since this method is used to calculate the best height for an image, we always return
+                // None for images.
+                AtomicKind::Empty | AtomicKind::Image(_) => None,
+            }
+        })
+    }
+
+    /// Turn this into a [`SizedAtomic`].
+    pub fn into_sized(
+        self,
+        ui: &Ui,
+        available_size: Vec2,
+        font_size: f32,
+        wrap_mode: Option<TextWrapMode>,
+    ) -> SizedAtomic<'a> {
+        let (preferred, kind) = self
+            .kind
+            .into_sized(ui, available_size, font_size, wrap_mode);
+        SizedAtomic {
+            size: self.size.unwrap_or_else(|| kind.size()),
+            preferred_size: preferred,
+            grow: self.grow,
+            kind,
+        }
+    }
+}
+
+/// A trait for conveniently building [`Atomic`]s.
+pub trait AtomicExt<'a> {
+    /// Set the atomic to a fixed size.
+    fn a_size(self, size: Vec2) -> Atomic<'a>;
+
+    /// Grow this atomic to the available space.
+    fn a_grow(self, grow: bool) -> Atomic<'a>;
+
+    /// Shrink this atomic if there isn't enough space.
+    ///
+    /// NOTE: Only a single [`Atomic`] may shrink for each widget.
+    fn a_shrink(self, shrink: bool) -> Atomic<'a>;
+}
+
+impl<'a, T> AtomicExt<'a> for T
+where
+    T: Into<Atomic<'a>> + Sized,
+{
+    fn a_size(self, size: Vec2) -> Atomic<'a> {
+        let mut atomic = self.into();
+        atomic.size = Some(size);
+        atomic
+    }
+
+    fn a_grow(self, grow: bool) -> Atomic<'a> {
+        let mut atomic = self.into();
+        atomic.grow = grow;
+        atomic
+    }
+
+    /// NOTE: Only a single atomic may be marked as shrink
+    fn a_shrink(self, shrink: bool) -> Atomic<'a> {
+        let mut atomic = self.into();
+        atomic.shrink = shrink;
+        atomic
+    }
+}
+
+impl<'a, T> From<T> for Atomic<'a>
+where
+    T: Into<AtomicKind<'a>>,
+{
+    fn from(value: T) -> Self {
+        Atomic {
+            size: None,
+            grow: false,
+            shrink: false,
+            kind: value.into(),
+        }
+    }
+}
+
+impl<'a> From<Image<'a>> for AtomicKind<'a> {
+    fn from(value: Image<'a>) -> Self {
+        AtomicKind::Image(value)
+    }
+}
+
+impl<'a, T> From<T> for AtomicKind<'a>
+where
+    T: Into<WidgetText>,
+{
+    fn from(value: T) -> Self {
+        AtomicKind::Text(value.into())
+    }
+}
+
+/// A list of [`Atomic`]s.
+#[derive(Clone, Debug, Default)]
+pub struct Atomics<'a>(Vec<Atomic<'a>>);
```

Comment:
Let's put each type into its own file under `egui/src/atomics/…` so that
A) Makes it easier to find the source code for each type
B) give each type room to grow
C) prevents a huge diff when we eventually decides to break up this file into a smaller files when it reached 2kloc or whatever :)

Also makes it easier to review (I can mark each type as reviewed when done)